src/main/java/graphql/execution/FieldCollector.java

```diff
@@ -41,12 +41,11 @@ public MergedSelectionSet collectFields(FieldCollectorParameters parameters, Mer
     public MergedSelectionSet collectFields(FieldCollectorParameters parameters, MergedField mergedField, boolean incrementalSupport) {
         Map<String, MergedField> subFields = new LinkedHashMap<>();
         Set<String> visitedFragments = new LinkedHashSet<>();
-        for (Field field : mergedField.getFields()) {
-            if (field.getSelectionSet() == null) {
-                continue;
+        mergedField.forEach(field -> {
+            if (field.getSelectionSet() != null) {
+                this.collectFields(parameters, field.getSelectionSet(), visitedFragments, subFields, null, incrementalSupport);
             }
-            this.collectFields(parameters, field.getSelectionSet(), visitedFragments, subFields, null, incrementalSupport);
-        }
+        });
```

Comment:
This was a consumer pattern - not wont materialise the list to consume the values